src/blocks/battery/upower.rs

```diff
@@ -57,86 +56,176 @@ impl Device {
+            None => Ok(None),
+            Some(device_connection) => {
+                let capacity = device_connection
+                    .device_proxy
+                    .percentage()
+                    .await
+                    .error("Failed to get capacity")?;
 
-        let time_remaining = match status {
-            BatteryStatus::Charging => Some(
-                self.device_proxy
-                    .time_to_full()
+                let power = device_connection
+                    .device_proxy
+                    .energy_rate()
                     .await
-                    .error("Failed to get time to full")? as f64,
-            ),
-            BatteryStatus::Discharging => Some(
-                self.device_proxy
-                    .time_to_empty()
+                    .error("Failed to get power")?;
+
+                let status = match device_connection
+                    .device_proxy
+                    .state()
                     .await
-                    .error("Failed to get time to empty")? as f64,
-            ),
-            _ => None,
-        };
+                    .error("Failed to get status")?
+                {
+                    1 => BatteryStatus::Charging,
+                    2 | 6 => BatteryStatus::Discharging,
+                    3 => BatteryStatus::Empty,
+                    4 => BatteryStatus::Full,
+                    5 => BatteryStatus::NotCharging,
+                    _ => BatteryStatus::Unknown,
+                };
+
+                let time_remaining = match status {
+                    BatteryStatus::Charging => Some(
+                        device_connection
+                            .device_proxy
+                            .time_to_full()
+                            .await
+                            .error("Failed to get time to full")? as f64,
+                    ),
+                    BatteryStatus::Discharging => Some(
+                        device_connection
+                            .device_proxy
+                            .time_to_empty()
+                            .await
+                            .error("Failed to get time to empty")? as f64,
+                    ),
+                    _ => None,
+                };
 
-        Ok(Some(BatteryInfo {
-            status,
-            capacity,
-            power: Some(power),
-            time_remaining,
-        }))
+                Ok(Some(BatteryInfo {
+                    status,
+                    capacity,
+                    power: Some(power),
+                    time_remaining,
+                }))
+            }
+        }
     }
 
     async fn wait_for_change(&mut self) -> Result<()> {
-        self.changes.next().await;
+        match &mut self.device_conn {
+            Some(device_connection) => {
+                select! {
+                    _ = device_connection.changes.next() => {},
+                    Some(msg)= self.device_removed_stream.next() => {
```

Comment:
I think we should also call `self.device_added_stream.next()` and discard its result, although I am not sure what happens if `MessageStream` runs out of its buffer space. Similarly `self.device_removed_stream.next()` should be called even if `self.device_conn` is `None`.